src/Microsoft.Identity.Web/TokenCacheProviders/Distributed/MsalDistributedTokenCacheAdapter.cs

```diff
@@ -4,42 +4,57 @@
+    /// An implementation of the token cache for both Confidential and Public clients backed by a Distributed Cache.
+    /// The Distributed Cache (L2), by default creates a Memory Cache (L1), for faster look up, resulting in a two level cache.
     /// </summary>
     /// <seealso>https://aka.ms/msal-net-token-cache-serialization</seealso>
     public class MsalDistributedTokenCacheAdapter : MsalAbstractTokenCacheProvider
     {
         /// <summary>
         /// .NET Core Memory cache.
         /// </summary>
-        private readonly IDistributedCache _distributedCache;
+        internal /*for tests*/ readonly IDistributedCache _distributedCache;
+        internal /*for tests*/ readonly MemoryCache _memoryCache;
+        private readonly ILogger<MsalDistributedTokenCacheAdapter> _logger;
+        private readonly TimeSpan? _expirationTime;
 
         /// <summary>
-        /// MSAL memory token cache options.
+        /// MSAL distributed token cache options.
         /// </summary>
-        private readonly MsalDistributedTokenCacheAdapterOptions _cacheOptions;
+        private readonly MsalDistributedTokenCacheAdapterOptions _distributedCacheOptions;
 
         /// <summary>
         /// Initializes a new instance of the <see cref="MsalDistributedTokenCacheAdapter"/> class.
         /// </summary>
-        /// <param name="memoryCache">Distributed cache instance to use.</param>
-        /// <param name="cacheOptions">Options for the token cache.</param>
+        /// <param name="distributedCache">Distributed cache instance to use.</param>
+        /// <param name="distributedCacheOptions">Options for the token cache.</param>
+        /// <param name="logger">MsalDistributedTokenCacheAdapter logger.</param>
         public MsalDistributedTokenCacheAdapter(
-                                            IDistributedCache memoryCache,
-                                            IOptions<MsalDistributedTokenCacheAdapterOptions> cacheOptions)
+                                            IDistributedCache distributedCache,
+                                            IOptions<MsalDistributedTokenCacheAdapterOptions> distributedCacheOptions,
+                                            ILogger<MsalDistributedTokenCacheAdapter> logger)
         {
-            if (cacheOptions == null)
+            if (distributedCacheOptions == null)
             {
-                throw new ArgumentNullException(nameof(cacheOptions));
+                throw new ArgumentNullException(nameof(distributedCacheOptions));
             }
 
-            _distributedCache = memoryCache;
-            _cacheOptions = cacheOptions.Value;
+            _distributedCache = distributedCache;
+            _distributedCacheOptions = distributedCacheOptions.Value;
+            _memoryCache = new MemoryCache(new MemoryCacheOptions { SizeLimit = _distributedCacheOptions.L1CacheSizeLimit * (1024 * 1024) });
```

Comment:
Is there a scenario where the user would want to overwrite this? Magic numbers and all...

Reply:
Yes, @bgavrilMS : 
we'd want the customer/services to decide which amount of memory they want to use for the L1 cache. By default this is 500Mb, but it's up to them to decide. It's a design decision.
See the MsalDistributedTokenCacheAdapterOptions properties that provides options which seem to make sense to avoid the L1 cache to grow forever (otherwise it would not be useful to have the L2 and all the memory will be exhausted)

Reply:
What @jmprieur said. we are trying to keep a lid on the memory growth, but allow it to be configurable.

---
In reply to: [586402386](https://github.com/AzureAD/microsoft-identity-web/pull/1037#discussion_r586402386) [](ancestors = 586402386)